src/driver/hermes_cdev.c

```diff
@@ -28,18 +28,199 @@
+		loff_t *f_pos)
+{
+	struct hermes_env *env = filp->private_data;
+	struct hermes_pci_dev *hpdev = env->hermes->hpdev;
+	struct xdma_channel *chnl;
+	long res;
+	loff_t pos;
+
+	chnl = xdma_get_c2h(hpdev);
+	if (!chnl) {
+		pr_err("No DMA channel available");
+		return -EIO;
+	}
+
+	if (env->data_slot < 0)
+		return -ENODATA;
+
+	if (count > hpdev->cfg.ehdssze) {
+		pr_warn("Data size greater than data slot size: 0x%lx > 0x%x. Resizing.\n",
+				count, hpdev->cfg.ehdssze);
+		count = hpdev->cfg.ehdssze;
+	}
+
+	pos = hpdev->cfg.ehdsoff + env->data_slot * hpdev->cfg.ehdssze;
+
+	res = xdma_channel_read_write(chnl, (char *) buff, count, &pos, false);
+
+	return res;
+}
+
+static ssize_t hermes_write(struct file *filp, const char __user *buff,
+		size_t count, loff_t *f_pos)
+{
+	struct hermes_env *env = filp->private_data;
+	struct hermes_pci_dev *hpdev = env->hermes->hpdev;
+	struct xdma_channel *chnl;
+	long res;
+	loff_t pos;
+
+	if (env->prog_slot < 0) {
+		pr_err("Program has not been downloaded to device. Aborting.\n");
+		return -ENOEXEC;
+	}
+
+	chnl = xdma_get_h2c(hpdev);
+	if (!chnl) {
+		pr_err("No DMA channel available");
+		return -EIO;
+	}
+
+	if (count > hpdev->cfg.ehdssze) {
+		pr_err("Data size greater than data slot size: 0x%lx > 0x%x\n",
+				count, hpdev->cfg.ehdssze);
+		return -EFBIG;
+	}
+
+	if (env->data_slot < 0) {
+		env->data_slot = hermes_request_slot(env, HERMES_SLOT_DATA);
+		if (env->data_slot < 0)
+			return env->data_slot;
+	}
+
+	pos = hpdev->cfg.ehdsoff + env->data_slot * hpdev->cfg.ehdssze;
+
+	res = xdma_channel_read_write(chnl, (char *) buff, count, &pos, true);
+
+	return res;
+}
+
+static long hermes_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+	struct hermes_env *env = filp->private_data;
+	struct hermes_pci_dev *hpdev = env->hermes->hpdev;
+	struct hermes_ioctl_argp *argp = (struct hermes_ioctl_argp *) arg;
+	struct xdma_channel *chnl;
+	long res;
+	loff_t pos;
+
+	chnl = xdma_get_h2c(hpdev);
+	if (!chnl) {
+		pr_err("No DMA channel available");
+		return -EIO;
+	}
+
+	if (argp->len > hpdev->cfg.ehpssze) {
+		pr_err("Program size greater than program slot size: 0x%lx > 0x%x\n",
+				argp->len, hpdev->cfg.ehpssze);
+		return -EFBIG;
+	}
+
+	if (env->prog_slot >= 0) {
+		pr_warn("Program was already downloaded, overwritting.\n");
+	} else {
+		env->prog_slot = hermes_request_slot(env, HERMES_SLOT_PROG);
+		if (env->prog_slot < 0)
+			return env->prog_slot;
+	}
+
+	pos = hpdev->cfg.ehpsoff + env->prog_slot * hpdev->cfg.ehpssze;
+
+	res = xdma_channel_read_write(chnl, (char *) argp->prog, argp->len,
+				      &pos, true);
+
+	if (res < 0)
+		return res;
+	else if (res != argp->len)
+		return -EIO;
+
+	return 0;
+}
+
 static const struct file_operations hermes_fops = {
 	.owner = THIS_MODULE,
+	.open = hermes_open,
+	.release = hermes_close,
+	.unlocked_ioctl = hermes_ioctl,
+	.read = hermes_read,
+	.write = hermes_write,
```

Comment:
We should really be implementing read_iter and write_iter.... It will give us a lot more flexibility with flags, etc.

Reply:
Aren't read_iter and write_iter for asynchronous I/O? The DMA is synchronous...

Reply:
No road_iter and write_iter are just a more general interface... They are necessary if you want to do async, but they are also necessary if you want to handle RWF_ flags in pwritev2/preadv2/io_uring. The flags might be a good way to trigger the program run.